addons/web_editor/models/field_html_history.py

```diff
@@ -0,0 +1,191 @@
+
+        if tag_prefix in ['+', 'R']:
+            for line_text in text_diff:
+                current_content_list.insert(line_id, line_text)
+        if tag_prefix in ['-', 'R']:
+            del current_content_list[line_index]
+
+    return '<'.join(current_content_list)
+
+# ------------------------------------------------------------
+# Custom difflib Diff methods
+# ------------------------------------------------------------
+
+
+def _format_range_context(start, stop):
+    beginning = start + 1  # lines start numbering with one
+    length = stop - start
+    if not length:
+        beginning -= 1  # empty ranges begin at line just before the range
+    if length <= 1:
+        return '{}'.format(beginning)
+    return '{},{}'.format(beginning, beginning + length - 1)
+
+
+def _custom_diff(new_content_str, old_content_str):
+    new_content_lines = new_content_str.split('<')
+    old_content_lines = old_content_str.split('<')
+
+    prefix = dict(insert='+', delete='-', replace='R', equal='=')
+    for group in SequenceMatcher(None, new_content_lines, old_content_lines, False).get_grouped_opcodes(0):
+        diff_lines = []
+        first, last = group[0], group[-1]
+        diff_string = '|'
+
+        diff_string += _format_range_context(first[1], last[2])
+        if any(tag in {'replace', 'delete'} for tag, _, _, _, _ in group):
+            for tag, i1, i2, _, _ in group:
+                if tag not in {'insert', 'equal', 'replace'}:
+                    diff_string = prefix[tag] + diff_string
+        if any(tag in {'replace', 'insert'} for tag, _, _, _, _ in group):
+            for tag, _, _, j1, j2 in group:
+                if tag not in {'delete', 'equal'}:
+                    diff_string = prefix[tag] + diff_string
+                    for line in old_content_lines[j1:j2]:
+                        diff_lines.append(line)
+        if diff_lines:
+            yield str(diff_string) + '<' + '<'.join(diff_lines)
+        else:
+            yield str(diff_string)
+
+
+class HtmlHistory(models.Model):
+    _name = "field.html.history"
+    _description = "Field html History"
+
+    history_diff_ids = fields.One2many(
+        "field.html.history.diff", "related_id", string="Body History",
+        copy=True)
+
+    def write(self, vals):
+        history_field_name = getattr(self, 'HTML_HISTORY_FIELD_NAME')
+        if history_field_name and history_field_name in vals:
+            new_body_str = vals[history_field_name]
+            """ Each change of body we need to create a diff for the history """
+            if self.id and isinstance(new_body_str, str):
+                old_body_str = getattr(self, history_field_name)
+                if isinstance(old_body_str, str) and new_body_str != old_body_str:
+                    diff = self.env['field.html.history.diff'].get_diff(new_body_str, old_body_str)
+                    self.env['field.html.history.diff'].create({"related_id": self.id,
```

Comment:
Try to be batch-enabled. Otherwise you are breaking the performances of any batched write ...

Reply:
I'm not familiar with batches in model as it's not often I work inside python Models. But I'll look into it to improve the write ;)